packages/analytics/dashboard-renderer/src/components/QueryDataProvider.vue

```diff
@@ -0,0 +1,79 @@
+<template>
+  <!-- Loading state -->
+  <KSkeleton
+    v-if="isLoading || (!data && !hasError)"
+    class="chart-skeleton"
+    type="table"
+  />
+  <!-- Error state -->
+  <KEmptyState
+    v-else-if="!isLoading && hasError"
+    cta-is-hidden
+    data-testid="chart-empty-state"
+    is-error
+  >
+    <template #message>
+      {{ errorMessage }}
+    </template>
+  </KEmptyState>
+
+  <slot
+    v-else
+    :data="data"
+  />
+</template>
+<script setup lang="ts">
+import { computed, inject, onUnmounted, ref } from 'vue'
+import useSWRV from 'swrv'
+import type { QueryProvider } from '../types/query-provider'
+import { INJECT_QUERY_PROVIDER } from '../types/query-provider'
+import { useSwrvState } from '@kong-ui-public/core'
+
+const props = defineProps<{
+  query: any // TODO: Explore v4
+  queryReady: boolean,
+}>()
+
+const emit = defineEmits(['queryComplete'])
+
+const queryBridge: QueryProvider | undefined = inject(INJECT_QUERY_PROVIDER)
+
+const queryKey = () => {
+  // SWRV can't accept `false` as a key, so use a more complicated conditional to match its `IKey` interface.
+  if (props.queryReady && queryBridge) {
+    return JSON.stringify(props.query)
+  }
+
+  return null
+}
+
+// Ensure that any pending requests are canceled on unmount.
+const abortController = new AbortController()
+
+onUnmounted(() => {
+  abortController.abort()
+})
+
+const { data, error, isValidating } = useSWRV(queryKey, async () => {
+  try {
+    // Note that queryBridge is guaranteed to be set here because SWRV won't execute the query if the key is null.
+    return queryBridge?.query(props.query, abortController)
+  } catch (e) {
+    // Note: 'Range not allowed' is defined by the API, therefore cannot be stored as string translation
+    // errorMessage.value = e?.response?.data?.message === 'Range not allowed for this tier'
+    //     ? i18n.t('configuration.vitals.reports.timeRangeExceeded')
+    //     : e?.response?.data?.message || e?.message
+    // TODO: Re-implement this once translations are in ^^
+    errorMessage.value = 'Error'
```

Comment:
question: is this error message useful? 

Regardless, it should be in `en.json`

Reply:
Since I needed to pull in translations anyway, I went ahead and implemented the original version.